perf-event/tests/breakpoint.rs

```diff
@@ -0,0 +1,64 @@
+use perf_event::{events, Builder};
+use std::fs::OpenOptions;
+use std::io::Write;
+
+// Need a function that will not be optimized away or inlined by the compiler.
+#[inline(never)]
+fn copy_and_fwrite(data: &[u8], file: &mut std::fs::File) -> std::io::Result<()> {
+    // This guarantees that the memory within data is read so that we can test
+    // read breakpoints.
+    let copy = data.to_vec();
+
+    file.write(&copy)?;
+    Ok(())
+}
+
+#[test]
+fn data() {
+    let mut file = OpenOptions::new()
+        .write(true)
+        .open("/dev/null")
+        .expect("Unable to open /dev/null for writing");
+    let data = b"TEST DATA".to_vec();
+
+    let mut counter = Builder::new()
+        .kind(events::Breakpoint::read_write(
+            data.as_ptr() as usize as _,
+            1,
+        ))
+        .observe_self()
+        .build()
+        .expect("Unable to build performance counter");
+    counter.enable().unwrap();
+
+    for _ in 0..1000 {
+        copy_and_fwrite(&data, &mut file).unwrap();
+    }
```

Comment:
For the data test, at least, let's have a function that uses `read_volatile` on a raw pointer. The compiler is (technically) free to change this to perform a single read to copy the data elsewhere and then do all the writes from the copy. But volatile reads are intended to be used when the access provokes some sort of hardware interaction - which seems like a pretty good fit for this test.

Reply:
I had actually forgotten that `read_volatile` existed. I've updated it and verified that the test works on my machine in both debug and release mode.